### PR TITLE
Clean and document symex-dereference

### PR DESCRIPTION
**Pull request by smowton**
Review commit-by-commit. As you can see there were various bits of goto-symex where I pulled the threads to confirm what they did, and found the threads didn't go anywhere.

### CODE REVIEW

**Review by peterschrammel — APPROVED** [review]

File: src/pointer-analysis/value_set_dereference.h [hunk 1]
```diff
@@ -50,19 +50,14 @@ class value_set_dereferencet
...
 
-  enum class modet { READ, WRITE };
-
   /// Dereference the given pointer-expression. Any errors are
   /// reported to the callback method given in the constructor.
   /// \param pointer: A pointer-typed expression, to
   ///        be dereferenced.
   /// \param guard: A guard, which is assumed to hold when dereferencing.
   /// \param mode: Indicates whether the dereferencing is a load or store
   //    (unused).
```
Comment (tautschnig) [inline src/pointer-analysis/value_set_dereference.h]: Documentation needs to be cleaned up as well.

File: src/goto-symex/symex_dereference_state.cpp [hunk 2]
```diff
@@ -13,6 +13,21 @@ Author: Daniel Kroening, kroening@kroening.com
...
+///
+/// For example, if \p expr is `p` and it has an `ID_C_failed_symbol` `p$object`
+/// (the naming convention used by `add_failed_symbols`), and the latest L1
+/// renaming of `p` is `p!2@4`, then we will create `p$object!2@4` if it doesn't
+/// already exist.
+///
+/// \param expr: expression to get or create a failed symbol for
+/// \param [out] symbol: failed symbol result
+/// \return true if we populated \p symbol
 bool symex_dereference_statet::get_or_create_failed_symbol(
```
Comment (tautschnig) [inline src/goto-symex/symex_dereference_state.cpp]: Not sure whether in this PR or in a separate one, but could we please change the return value to `optionalt<symbolt>`? The "Boolean return value + out-parameter" combination is just too error-prone and nowadays we have better options.

**Review by tautschnig — APPROVED** [review]

File: src/goto-symex/symex_dereference_state.cpp (hunk 2 above)
Comment (smowton) [reply src/goto-symex/symex_dereference_state.cpp]: Pushed an extra commit to just use `symbolt *`, since that's what the existing code was expecting

**Comment by tautschnig** [comment]
This now needs a rebase (because I just merged the codet API PR) - sorry for the extra work!

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 115cb7eaeda55e8c2269398e4a89eefb9e420e0f).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/100604500

**Pull request merged**